Cards/CardBase.gd

```diff
@@ -31,29 +31,49 @@ func parse_card_data(card_data: Dictionary) -> void:
 	# TODO
 	pass
 
-func _apply_all_effects(target: Entity) -> void:
-	for effect_data: EffectData in card_effects_data:
-		effect_data.apply_effect_data(target)
+func _apply_all_effects(targets: Array[Entity], effects: Array[EffectData]) -> void:
+	for target : Entity in targets:
+		for effect : EffectData in effects:
+			effect.apply_effect_data(target)
 
 func can_play_card(caster: Entity, target: Entity) -> bool:
 	return caster.get_party_component().can_play_on_entity(application_type, target)
 
 
 func on_card_play(caster: Entity, targets: Array[Entity]) -> void:
 	#Split up targeted attacks and all attacks
+	var target_effects: Array[EffectData] = get_target_effects()
+	var all_effects: Array[EffectData] = get_all_effects()
+	
+	#Apply single target
+	_apply_all_effects(targets, target_effects)
+	#Get every unit that is to be affected by card
+	var all_target : Array[Entity] = get_all_targets()
+
+	#apply effect to every target
+	_apply_all_effects(all_target, all_effects)
+	
+	CardManager.on_card_action_finished.emit(self)
+
+func get_target_effects() -> Array[EffectData]:
 	var target_effects: Array[EffectData] = []
-	var all_effects: Array[EffectData] = []
 	
 	for effect_data: EffectData in card_effects_data:
 		if(effect_data.target_type == Enums.TargetType.SINGLE_TARGET):
-			target_effects.push_front(effect_data)
-		else :
-			all_effects.push_front( effect_data)
-	#Apply single target
-	for entity : Entity in targets:
-		for effect : EffectData in target_effects:
-			effect.apply_effect_data(entity)
-	#Get every unit that is to be affected by card
+			target_effects.append(effect_data)
+			
+	return target_effects
+
+func get_all_effects() -> Array[EffectData]:
```

Comment:
the name is a bit ambiguous. I would expect it to mean "all the effects of the card" but you mean it as "all the effects that apply to a target ALL"
`get_effects_target_all` would probably be a clearer name

Reply:
I can change that

Reply:
@JonaLam unresolving this because i see it's back to the name I had a problem with